server/src/main/java/org/candlepin/model/AbstractHibernateCurator.java

```diff
@@ -767,47 +770,205 @@ protected int safeSQLUpdateWithCollection(String sql, Collection<?> collection,
+     */
+    public void lock(E entity, LockModeType lockMode) {
+        this.getEntityManager().lock(entity, lockMode);
+    }
 
-        if (ids != null && ids.iterator().hasNext()) {
-            StringBuilder hql = new StringBuilder("SELECT obj FROM ")
-                .append(entityName)
-                .append(" obj WHERE ")
-                .append(keyName)
-                .append(" IN (:ids)");
+    /**
+     * Refreshes/reloads the given entity and locks it using a pessimistic write lock.
+     *
+     * @param entity
+     *  The entity to lock and load
+     *
+     * @return
+     *  The locked entity
+     */
+    public E lockAndLoad(E entity) {
+        this.getEntityManager().refresh(entity, LockModeType.PESSIMISTIC_WRITE);
 
-            javax.persistence.Query query = this.getEntityManager()
-                .createQuery(hql.toString())
-                .setLockMode(LockModeType.PESSIMISTIC_WRITE);
+        return entity;
+    }
 
-            for (List<String> block : Iterables.partition(ids, IN_OPERATOR_BLOCK_SIZE)) {
-                query.setParameter("ids", block);
-                result.addAll((List<E>) query.getResultList());
-            }
+    /**
+     * Locks the given collection of entities, reloading them as necessary.
+     *
+     * @param entities
+     *  A collection of entities to lock
+     *
+     * @throws RuntimeException
+     *  If this method is called for a curator handling an entity type that is not a subclass of
+     *  the AbstractHibernateObject class.
+     *
+     * @return
+     *  The collection of locked entities
+     */
+    public Iterable<E> lockAndLoad(Iterable<E> entities) {
+        // Impl note:
+        // We're going to take advantage of some blackbox knowledge of how LockAndLoadByIds works to
+        // minimize the amount of extra loops we need to do. We can pass a custom iterable which
+        // fetches the entity's ID on the call to "next" and pass that through to LockAndLoadByIds.
+        //
+        // We'll need to duplicate this logic for anything that isn't a Persisted instance or uses
+        // a different name for their primary key (see: products and content).
+
+        if (!AbstractHibernateObject.class.isAssignableFrom(this.entityType)) {
+            throw new RuntimeException("the lockAndLoad method does not support " +
+                this.entityType.getSimpleName() + " entities.");
+        }
 
-            //In some situations, even after locking the entity we
-            //got stale in the entity e.g. Pool.consumed
-            //This refresh reloads the entity after the lock has
-            //been issued.
-            for (E e : result) {
-                getEntityManager().refresh(e);
-            }
+        if (entities == null) {
+            return Collections.<E>emptyList();
         }
 
-        return result;
+        // We redeclare the collection here so we don't require the final modifier in subclass
```

Comment:
The bulk of this change was started because there's a need to add sorting to our bulk locking operations to avoid some deadlock situations. This method offloads the bulk of its work to the ```lockAndLoadByColumn``` method, which we know (through blackbox knowledge) is going to put the entities in a sorted set via loop if the iterable it receives isn't a Collection. So, rather than loop through the entities there and pull the IDs, we wrap the whole thing in the anon iterable and do the sorting and ID pulling in one loop, rather than two.